### ISSUE TITLE
Systematize the error-prone stdenv bootstrapping dance

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#11486]
Currently, both the linux and darwin pure stdenvs need to jump through some careful error-prone hoops of build stages with different sets of `allowedRequisites` to end up with a stdenv that has no runtime dependencies on the bootstrap bundle.

We should be able to automate the logic for building that sequence of bootstrap steps by enforcing that all packages used during a stdenv build have `allowedRequisites` (or perhaps a new Nix flag called `expectedRequisites` that is supposed to be an exact bound) specified, and solving the constraints on the fly. I haven't written the actual algorithm yet, but don't expect it to be particularly computationally expensive, and it could save lots of effort when bootstrapping a new pure stdenv or maintaining one of the existing ones.

Can anyone see obvious problems with this, assuming the computation isn't too expensive?

cc @shlevy @edolstra @pikajude @vcunat who I think have dealt quite a bit with low-level stdenv stuff.

**Comment by shlevy** [comment]
No obvious issues stand out to me.

**Comment by Profpatsch** [comment]
(triage) @copumpkin Do you plan to implement it?

**Comment by copumpkin** [comment]
@Profpatsch yes

**Comment by Ericson2314** [comment]
I beleive my follow-up on my PR above would do this "constructively". The basic idea is to have a runtime package set, and buildtime package set, which callPackage splices together for backwards compat.

For cross compiling you have: last normal stage->cross compile build stage with build = host != target, and then run stage with build != host = target. Note how this correctly makes a build dependency of a build dependency built normally.

For bootstrapping:

Phase0Build is bootstrap tarball

PhaseN+1Build is overridden with exactly the desired build deps from the previous phase, and fixed with itself as *it's* build package set.

PhaseN uses phaseNBuild as build packages, si the previous phase is purged at runtime.

I'll wrote a code quote when I am off phone.

**Comment by copumpkin** [comment]
Not sure I understand. What would a bootstrap look like in your model? My main goal is to avoid having to figure it all out again when something grows a new dependency somewhere in the middle.

**Comment by Ericson2314** [comment]
Oh I see you mean alloweeRequisites per package. What I said is probably orthogonal then.

**Comment by copumpkin** [comment]
Well, it's more than that. It's the full set of bootstrap steps that led to the darwin bootstrap taking 5 or 6 steps and the linux one taking a different number of steps, and so on. It took me a lot of staring at `nix-store -q --graph` output in graphviz to figure out how to get a pure bootstrap, and I don't want myself or anyone else to have to keep doing that when they/we make changes to the stdenv.

**Comment by copumpkin** [comment]
I _think_ the process that leads to those steps should be automatable, if you give Nix the right information to reason about the steps programmatically.

**Comment by Ericson2314** [comment]
@copumpkin I'd be interested in helping do this now, now that https://github.com/NixOS/nixpkgs/issues/15043 is finally done.

**Comment by copumpkin** [comment]
Do you have a rough picture of what is needed to make this happen? I was hoping to do it but as you can see, it's been hard to carve out time for it.

**Comment by Ericson2314** [comment]
Not entirely. I get that the primarily compiler (gcc, clang) can only be bootstrapped in so few steps, and I get trying to make sure nothing keeps a runtime dependency on the bootstrap archive (or its contents). I do not have intuition for the difficulty of the latter---shouldn't most those runtime dependencies just not exist in the first place?

**Comment by Profpatsch** [comment]
(triage) has this been resolved?

**Comment by Ericson2314** [comment]
No. But now that LLVM cross support is fairly close, we are almost ready to do this.
